<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from drug_robot.xacro               | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="drug_robot">
  <!-- DRUG DIFFERENTIAL DRIVE XACRO -->
  <!-- in kg-->
  <!-- HOKUYO LASER SCANNER -->
  <!-- CAMERA -->
  <!-- in kg-->
  <!-- LINK -->
  <link name="drug_base_footprint">
    </link>
  <joint name="drug_base_footprint_joint" type="fixed">
    <parent link="drug_base_footprint"/>
    <child link="drug_base_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <!-- BODY -->
  <link name="drug_base_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="white">
        <color rgba="1 1 1 1"/>
      </material>
      <geometry>
        <cylinder length="0.2" radius="0.375"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.2" radius="0.375"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="5"/>
      <inertia ixx="0.1924479166666667" ixy="0" ixz="0" iyy="0.1924479166666667" iyz="0" izz="0.3515625"/>
    </inertial>
  </link>
  <gazebo reference="drug_base_link">
    <material>Gazebo/White</material>
  </gazebo>
  <!-- BOX CYLINDER BODY -->
  <link name="box_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="yellow">
        <color rgba="1 1 0 1"/>
      </material>
      <geometry>
        <cylinder length="0.5" radius="0.3"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.5" radius="0.3"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2"/>
      <inertia ixx="0.08666666666666667" ixy="0" ixz="0" iyy="0.08666666666666667" iyz="0" izz="0.09"/>
    </inertial>
  </link>
  <joint name="box_joint" type="fixed">
    <parent link="drug_base_link"/>
    <child link="box_link"/>
    <origin rpy="0 0 0" xyz="0 0 0.35"/>
  </joint>
  <gazebo reference="box_link">
    <material>Gazebo/YellowGlow</material>
  </gazebo>
  <!-- lr: left, right -->
  <link name="drug_right_wheel">
    <visual>
      <origin rpy="1.5707963267948966 0  0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.06" radius="0.11"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.06" radius="0.11"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.008312499999999999" ixy="0" ixz="0" iyy="0.008312499999999999" iyz="0" izz="0.015125000000000001"/>
    </inertial>
  </link>
  <gazebo reference="drug_right_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
    <material>Gazebo/Black</material>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <joint name="drug_right_wheel_joint" type="continuous">
    <parent link="drug_base_link"/>
    <child link="drug_right_wheel"/>
    <origin rpy="0 0 0" xyz="0.0 -0.375 -0.04"/>
    <axis rpy="0  0" xyz="0 1 0"/>
    <limit effort="100" velocity="50"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>
  <!-- Transmission is important to link the joints and the controller -->
  <transmission name="drug_right_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="drug_right_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
    <actuator name="drug_right_wheel_joint_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- lr: left, right -->
  <link name="drug_left_wheel">
    <visual>
      <origin rpy="1.5707963267948966 0  0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.06" radius="0.11"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5707963267948966 0 0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.06" radius="0.11"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.008312499999999999" ixy="0" ixz="0" iyy="0.008312499999999999" iyz="0" izz="0.015125000000000001"/>
    </inertial>
  </link>
  <gazebo reference="drug_left_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
    <material>Gazebo/Black</material>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <joint name="drug_left_wheel_joint" type="continuous">
    <parent link="drug_base_link"/>
    <child link="drug_left_wheel"/>
    <origin rpy="0 0 0" xyz="0.0 0.375 -0.04"/>
    <axis rpy="0  0" xyz="0 1 0"/>
    <limit effort="100" velocity="50"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>
  <!-- Transmission is important to link the joints and the controller -->
  <transmission name="drug_left_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="drug_left_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
    <actuator name="drug_left_wheel_joint_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- CASTER FRONT WHEEL-->
  <link name="drug_caster_front_wheel">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="grey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
      <geometry>
        <sphere radius="0.06"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.06"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>
  <joint name="drug_caster_front_wheel_joint" type="fixed">
    <parent link="drug_base_link"/>
    <child link="drug_caster_front_wheel"/>
    <origin rpy="0 0 0" xyz="0.3 0 -0.085"/>
  </joint>
  <gazebo reference="drug_caster_front_wheel">
    <material>Gazebo/GreyTransparent</material>
  </gazebo>
  <!-- CASTER BACK WHEEL-->
  <link name="drug_caster_back_wheel">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="grey">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
      <geometry>
        <sphere radius="0.06"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.06"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>
  <joint name="drug_caster_back_wheel_joint" type="fixed">
    <parent link="drug_base_link"/>
    <child link="drug_caster_back_wheel"/>
    <origin rpy="0 0 0" xyz="-0.3 0 -0.085"/>
  </joint>
  <gazebo reference="drug_caster_back_wheel">
    <material>Gazebo/GreyTransparent</material>
  </gazebo>
  <link name="drug_hokuyo_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="blue">
        <color rgba=" 0 0 1 1"/>
      </material>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
    </visual>
  </link>
  <joint name="drug_hokuyo_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.35 0 0.1125"/>
    <parent link="drug_base_link"/>
    <child link="drug_hokuyo_link"/>
  </joint>
  <gazebo reference="drug_hokuyo_link">
    <material>Gazebo/Blue</material>
    <turnGravityOff>false</turnGravityOff>
    <sensor name="head_hokuyo_sensor" type="ray">
      <pose>0.025 0 0 0 0 0</pose>
      <visualize>false</visualize>
      <update_rate>30</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>720</samples>
            <resolution>1</resolution>
            <min_angle>-1.570796</min_angle>
            <max_angle>1.570796</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.10</min>
          <max>10.0</max>
          <resolution>0.001</resolution>
        </range>
      </ray>
      <plugin filename="libgazebo_ros_laser.so" name="gazebo_ros_head_hokuyo_controller">
        <topicName>/drug/hokuyo_sensor/scan</topicName>
        <frameName>drug_hokuyo_link</frameName>
      </plugin>
    </sensor>
  </gazebo>
  <link name="drug_camera_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="orange">
        <color rgba="1 0.65 0 1"/>
      </material>
      <geometry>
        <box size="0.04 0.08 0.03"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.04 0.08 0.03"/>
      </geometry>
    </collision>
  </link>
  <joint name="drug_camera_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.355 0 0.1525"/>
    <parent link="drug_base_link"/>
    <child link="drug_camera_link"/>
  </joint>
  <gazebo reference="drug_camera_link">
    <material>Gazebo/Orange</material>
    <sensor name="realsense_R200" type="depth">
      <always_on>true</always_on>
      <visualize>false</visualize>
      <camera>
        <horizontal_fov>1.3439</horizontal_fov>
        <image>
          <width>1920</width>
          <height>1080</height>
          <format>R8G8B8</format>
        </image>
        <depth_camera/>
        <clip>
          <near>0.03</near>
          <far>100</far>
        </clip>
      </camera>
      <plugin filename="libgazebo_ros_openni_kinect.so" name="camera_controller">
        <baseline>0.2</baseline>
        <alwaysOn>true</alwaysOn>
        <updateRate>30.0</updateRate>
        <cameraName>camera</cameraName>
        <frameName>drug_camera_link</frameName>
        <imageTopicName>rgb/image_raw</imageTopicName>
        <depthImageTopicName>depth/image_raw</depthImageTopicName>
        <pointCloudTopicName>depth/points</pointCloudTopicName>
        <cameraInfoTopicName>rgb/camera_info</cameraInfoTopicName>
        <depthImageCameraInfoTopicName>depth/camera_info</depthImageCameraInfoTopicName>
        <pointCloudCutoff>0.4</pointCloudCutoff>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
        <CxPrime>0.0</CxPrime>
        <Cx>0.0</Cx>
        <Cy>0.0</Cy>
        <focalLength>0</focalLength>
        <hackBaseline>0</hackBaseline>
      </plugin>
    </sensor>
  </gazebo>
  <!-- Gazebo plugin for ROS Control -->
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
      <controlPeriod>0.001</controlPeriod>
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
      <legacyModeNS>true</legacyModeNS>
    </plugin>
  </gazebo>
  <!-- Gazebo plugin for differential drive controller -->
  <gazebo>
    <plugin filename="libgazebo_ros_diff_drive.so" name="differential_drive_controller">
      <legacyMode>true</legacyMode>
      <rosDebugLevel>na</rosDebugLevel>
      <alwaysOn>true</alwaysOn>
      <publishTf>true</publishTf>
      <broadcastTF>true</broadcastTF>
      <publishOdomTF>true</publishOdomTF>
      <publishWheelTF>false</publishWheelTF>
      <publishWheelJointState>false</publishWheelJointState>
      <wheelSeparation>0.75</wheelSeparation>
      <wheelDiameter>0.22</wheelDiameter>
      <updateRate>30.0</updateRate>
      <wheelTorque>30</wheelTorque>
      <wheelAcceleration>1.8</wheelAcceleration>
      <commandTopic>/cmd_vel</commandTopic>
      <odometryTopic>/odom</odometryTopic>
      <odometryFrame>drug_odom</odometryFrame>
      <odometrySource>world</odometrySource>
      <robotBaseFrame>drug_base_footprint</robotBaseFrame>
    </plugin>
  </gazebo>
</robot>

